video/vdu_layers.h

```diff
@@ -0,0 +1,1148 @@
+		default: {
+			debug_log("vdu_sys_layers_tilebank_free: Invalid tileBankNum %d specified.\r\n",tileBankNum);
+		}
+	}
+}
+
+void VDUStreamProcessor::vdu_sys_layers_tilemap_init(uint8_t tileLayerNum, uint8_t tileMapSize) {
+
+	// Check if tileMap0 already exists and free if it is.
+
+	switch (tileLayerNum) {
+
+		case 0: {		// If tilemap/layer number is 0
+			
+			// Check if already exists
+	
+			if (tileMap0 !=NULL) {
+				// If already exists, then free and reinitialise
+
+				vdu_sys_layers_tilemap_free(tileLayerNum); 
+			}
+			
+		} break;
+
+		default: {
+			debug_log("vdu_sys_layers_tilemap_init: Invalid tileLayerNum %d specified.\r\n",tileLayerNum);
+			return;
+		}
+	}
+
+	//	The following tile map sizes are supported:
+	//	0=32x32, 1=32x64, 2=32x128, 3=64x32, 4=64x64, 5=64x128, 6=128x32, 7=128x64, 8=128x128
+
+	switch (tileMapSize) {
+
+		case 0: {		// 32x32 tilemap
+			
+			tileMap0Properties.width = 32;
+			tileMap0Properties.height = 32;
+
+		} break;
+
+		case 1: {		// 32x64 tilemap
+			
+			tileMap0Properties.width = 32;
+			tileMap0Properties.height = 64;
+
+		} break;
+
+		case 2: {		// 32x128 tilemap
+			
+			tileMap0Properties.width = 32;
+			tileMap0Properties.height = 128;
+
+		} break;
+
+		case 3: {		// 64x32 tilemap
+			
+			tileMap0Properties.width = 64;
+			tileMap0Properties.height = 32;
+
+		} break;
+
+		case 4: {		// 64x64 tilemap
+			
+			tileMap0Properties.width = 64;
+			tileMap0Properties.height = 64;
+
+		} break;
+
+		case 5: {		// 64x128 tilemap
+			
+			tileMap0Properties.width = 64;
+			tileMap0Properties.height = 128;
+
+		} break;
+
+		case 6: {		// 128x32 tilemap
+			
+			tileMap0Properties.width = 128;
+			tileMap0Properties.height = 32;
+
+		} break;
+
+		case 7: {		// 128x64 tilemap
+			
+			tileMap0Properties.width = 128;
+			tileMap0Properties.height = 64;
+
+		} break;
+
+		case 8: {		// 128x128 tilemap
+			
+			tileMap0Properties.width = 128;
+			tileMap0Properties.height = 128;
+
+		} break;
+
+		default: {
+			debug_log("vdu_sys_layers_tilemap_init: Invalid tileMapSize %d specified.\r\n",tileMapSize);
+			return;
+		}
+	}
+
+	uint8_t tileMapWidth = tileMap0Properties.width;
+	uint8_t tileMapHeight = tileMap0Properties.height;
+
+	int tileMapBufferSize = tileMapWidth * sizeof(struct Tile*);
+
+	// Initial release to only support a single tile layer
+	if (tileLayerNum != 0) {
+		debug_log("vdu_sys_layers_tilemap_init: Invalid tileLayerNum %d specified.\r\n",tileLayerNum);
+		return;
+	}
+
+	switch (tileLayerNum) {
+
+		case 0: {		// If tile map number is 0
+						
+			// Allocate memory for each column in the tile map
+			tileMap0 = (struct Tile**)heap_caps_malloc(tileMapBufferSize,MALLOC_CAP_SPIRAM);
+
+			// Allocate memory for each row in the tile map
+			for (auto i=0; i<tileMapWidth; i++) {
+				tileMap0[i] = (struct Tile*)heap_caps_malloc(tileMapHeight * sizeof(struct Tile),MALLOC_CAP_SPIRAM);
+			}
+						
+			// Only continue if the init was successful
+			
+			if (tileMap0 != NULL) {
```

Comment:
I think there's a minor logical omission here

whilst the final line here checks to see if `timeMap0` was null, there's no detection for a failure of allocation for an individual row within the tilemap

```suggestion
			bool allocFail = false;

			// Allocate memory for each column in the tile map
			tileMap0 = (struct Tile**)heap_caps_malloc(tileMapBufferSize,MALLOC_CAP_SPIRAM);
			if (tileMap0 == nullptr) {
				allocFail = true;
			}

			// Allocate memory for each row in the tile map
			for (auto i=0; i<tileMapWidth; i++) {
				tileMap0[i] = (struct Tile*)heap_caps_malloc(tileMapHeight * sizeof(struct Tile),MALLOC_CAP_SPIRAM);
				if (!tileMap0[I]) {
					allocFail = true;
					break;
				}
			}
						
			// Only continue if the init was successful
			
			if (allocFail) {
```